PowerShellTools.HostService/ServiceManagement/Debugging/PowerShellDebuggingService.cs

```diff
@@ -204,80 +241,178 @@ public void AttachToRunspace(uint pid)
+                DebugScenario preScenario = GetDebugScenario();
+
+                _attachRequestEvent.Reset();
+                _currentPowerShell.Runspace = _runspace;
+                _currentPowerShell.AddCommand("Exit-PSHostProcess");
+                _currentPowerShell.Invoke();
+
+                // wait for invoke to finish swapping the runspaces if detaching from a local process
+                if (preScenario == DebugScenario.LocalAttach)
                 {
-                    if (module.ModuleName.Equals("powershell.exe", StringComparison.Ordinal) && _installedPowerShellVersion >= RequiredPowerShellVersionForProcessAttach)
+                    _attachRequestEvent.WaitOne(DebugEngineConstants.AttachRequestEventTimeout);
+
+                    // scenario after exiting, used to determine if semaphore timed out, should not match our pre-scenario
+                    DebugScenario postScenario = GetDebugScenario();
+
+                    // make sure that the semaphore didn't just time out
+                    if (preScenario == postScenario)
                     {
-                        return true;
+                        // very unlikely for this to happen, but we should make sure to handle the case anyway
+                        ServiceCommon.Log("Failed to detach from running process");
+                        return false;
                     }
                 }
+
+                // rehook event handlers
+                AddEventHandlers();
             }
-            catch (Exception ex)
+
+            return true;
+        }
+
+        /// <summary>
+        /// Finds all processes on a remote computer which have loaded the powershell.exe module
+        /// </summary>
+        /// <param name="remoteName">Name of the remote machine</param>
+        /// <returns></returns>
+        public List<KeyValuePair<uint, string>> EnumerateRemoteProcesses(string remoteName)
+        {
+            List<KeyValuePair<uint, string>> validProcesses = new List<KeyValuePair<uint, string>>();
+
+            // Retrieve callback context so credentials window can display
+            if (_callback == null)
             {
-                ServiceCommon.Log(string.Format("{0} , cannot examine modules of process; id: {1}", ex.Message, pid));
+                _callback = OperationContext.Current.GetCallbackChannel<IDebugEngineCallback>();
             }
-            return false;
+
+            using (_currentPowerShell = PowerShell.Create())
+            {
+                try
+                {
+                    // Initiate remote session with the remote machine
+                    _currentPowerShell.Runspace = _runspace;
+                    _currentPowerShell.AddScript(string.Format(DebugEngineConstants.EnterRemoteSessionDefaultCommand, remoteName));
+                    _currentPowerShell.Invoke();
+
+                    if (GetDebugScenario() == DebugScenario.Local)

```

Comment:
This case happens if the user hits "Cancel" on the login dialog or enters the wrong credentials. If you look in the method connect() you will see the effect of this.